File: windows/src/PrivateWhisper/CleanupService.cs
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace PrivateWhisper;

/// <summary>
/// Talks to an OpenAI-compatible /chat/completions endpoint (LM Studio or the
/// embedded llama-server) for the cleanup pass and command-mode rewrites.
/// Any failure throws — the pipeline falls back to the raw transcript so
/// dictation is never lost. Logic ported 1:1 from the Mac CleanupService;
/// the prompts come from the shared, eval-verified resources.
/// </summary>
public sealed class CleanupService
{
    private static readonly HttpClient Http = new() { Timeout = System.Threading.Timeout.InfiniteTimeSpan };

    public string BaseUrl { get; }
    public string Model { get; }
    public double TimeoutSeconds { get; }

    public CleanupService(string baseUrl, string model, double timeoutSeconds)
    {
        BaseUrl = baseUrl;
        Model = model;
        TimeoutSeconds = timeoutSeconds;
    }

    /// <summary>Canonical prompt from shared/prompts/cleanup_prompt.txt.</summary>
    public static readonly string SystemPrompt = EmbeddedResources.ReadText("cleanup_prompt.txt");

    /// <summary>Canonical prompt from shared/prompts/rewrite_prompt.txt.</summary>
    public static readonly string RewritePrompt = EmbeddedResources.ReadText("rewrite_prompt.txt");

    public sealed class CleanupException : Exception
    {
        public CleanupException(string message) : base(message) { }
    }

    public async Task<string> CleanupAsync(
        string transcript, string? language,
        IReadOnlyList<string> glossary, string? toneHint)
    {
        var systemPrompt = new StringBuilder(SystemPrompt);
        if (!string.IsNullOrEmpty(language) && language != "unknown")
        {
            systemPrompt.Append($"\n- The input language is \"{language}\". The output must be in that same language.");
        }
        if (glossary.Count > 0)
        {
            systemPrompt.Append("\n- Personal dictionary — when the transcript contains a similar-sounding or misspelled variant of one of these, use this exact spelling: ")
                .Append(string.Join(", ", glossary));
        }
        if (!string.IsNullOrEmpty(toneHint))
        {
            systemPrompt.Append($"\n- The text will be inserted into an app where the expected style is: {toneHint}. Adjust register lightly; never change meaning.");
        }

        // PRD: max_tokens sized to input length × 1.5 (≈3 chars/token heuristic),
        // plus fixed headroom for reasoning models that "think" before answering —
        // otherwise they hit the cap mid-thought and return empty content.
        // Non-thinking models just stop early, so the headroom is free.
        int maxTokens = Math.Max(256, (int)(transcript.Length / 3.0 * 1.5) + 64) + 2048;

        return await SendAsync(systemPrompt.ToString(), transcript, maxTokens);
    }

    /// <summary>Command mode: apply a spoken instruction to selected text.</summary>
    public async Task<string> RewriteAsync(string selection, string instruction)
    {
        string user = $"Instruction: {instruction}\n\nText:\n{selection}";
        int maxTokens = Math.Max(512, (int)(selection.Length / 3.0 * 2)) + 2048;
        return await SendAsync(RewritePrompt, user, maxTokens);
    }

    private async Task<string> SendAsync(string systemPrompt, string userContent, int maxTokens)
    {
        string url = BaseUrl.Trim('/') + "/chat/completions";

        // Qwen3 hybrids honor /no_think in the prompt; belt-and-braces next to
        // reasoning_effort below. Harmless elsewhere.
        if (Model.Contains("qwen", StringComparison.OrdinalIgnoreCase))
        {
            systemPrompt += "\n/no_think";
        }

        var body = new JsonObject
        {
            ["model"] = Model,
            ["messages"] = new JsonArray(
                new JsonObject { ["role"] = "system", ["content"] = systemPrompt },
                new JsonObject { ["role"] = "user", ["content"] = userContent }),
            ["temperature"] = 0.2,
            ["max_tokens"] = maxTokens,
            ["stream"] = false,
            // Disables thinking on models that support it (e.g. Qwen 3.5 —
            // 0.46s instead of 67s); safely ignored by non-reasoning models.
            ["reasoning_effort"] = "none",
        };

        string responseBody;
        try
        {
            using var content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(TimeoutSeconds));
            using HttpResponseMessage response = await Http.PostAsync(url, content, cts.Token);
            responseBody = await response.Content.ReadAsStringAsync(cts.Token);
            if (!response.IsSuccessStatusCode)
            {
                throw new CleanupException(
                    $"Cleanup backend error: HTTP {(int)response.StatusCode} {Truncate(responseBody, 300)}");
            }
        }
        catch (CleanupException)
        {
            throw;
        }
        catch (Exception ex)
        {
            throw new CleanupException("Cleanup request failed: " + ex.Message);
        }

        string? messageContent = null;
        try
        {
            using JsonDocument doc = JsonDocument.Parse(responseBody);
            if (doc.RootElement.TryGetProperty("choices", out JsonElement choices) &&
                choices.ValueKind == JsonValueKind.Array && choices.GetArrayLength() > 0 &&
                choices[0].TryGetProperty("message", out JsonElement message) &&
                message.TryGetProperty("content", out JsonElement contentEl))
            {
                messageContent = contentEl.GetString();
            }
        }
        catch (JsonException)
        {
            // handled below
        }
        if (messageContent == null)
        {
            throw new CleanupException("Cleanup backend returned an unexpected response shape");
        }

        string cleaned = StripReasoning(messageContent).Trim();
        if (cleaned.Length == 0)
        {
            throw new CleanupException("Cleanup backend returned an empty result");
        }
        return cleaned;
    }

    private static string Truncate(string s, int max) => s.Length <= max ? s : s[..max];

    /// <summary>Removes &lt;think&gt;…&lt;/think&gt; blocks that reasoning models may emit.</summary>
    public static string StripReasoning(string text)
    {
        string result = Regex.Replace(text, "<think>.*?</think>", "", RegexOptions.Singleline);
        // Unterminated think block (hit max_tokens mid-reasoning) → nothing usable.
        int open = result.IndexOf("<think>", StringComparison.Ordinal);
        if (open >= 0)
        {
            result = result[..open];
        }
        return result;
    }

    /// <summary>Fetches available model IDs from LM Studio (settings dropdown).</summary>
    public static async Task<IReadOnlyList<string>> AvailableModelsAsync(string baseUrl)
    {
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(3));
            using HttpResponseMessage response = await Http.GetAsync(baseUrl.Trim('/') + "/models", cts.Token);
            string body = await response.Content.ReadAsStringAsync(cts.Token);
            using JsonDocument doc = JsonDocument.Parse(body);
            if (doc.RootElement.TryGetProperty("data", out JsonElement data) &&
                data.ValueKind == JsonValueKind.Array)
            {
                var models = new List<string>();
                foreach (JsonElement item in data.EnumerateArray())
                {
                    if (item.TryGetProperty("id", out JsonElement id) && id.GetString() is string s)
                    {
                        models.Add(s);
                    }
                }
                return models;
            }
        }
        catch
        {
            // unreachable → empty list
        }
        return Array.Empty<string>();
    }

    // ---- Backend routing (port of the Mac resolveBackend) ----

    /// <summary>Quick reachability probe with a short cache (avoids a 1 s stall
    /// on every dictation when LM Studio is down).</summary>
    private static (string Url, bool Reachable, DateTime At)? lastProbe;
    private static readonly object ProbeLock = new();

    public static async Task<bool> IsReachableAsync(string baseUrl)
    {
        lock (ProbeLock)
        {
            if (lastProbe is { } probe && probe.Url == baseUrl &&
                (DateTime.UtcNow - probe.At).TotalSeconds < 30)
            {
                return probe.Reachable;
            }
        }
        bool reachable;
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
            using HttpResponseMessage response = await Http.GetAsync(baseUrl.Trim('/') + "/models", cts.Token);
            reachable = true; // any HTTP response counts, like the Mac probe
        }
        catch
        {
            reachable = false;
        }
        lock (ProbeLock)
        {
            lastProbe = (baseUrl, reachable, DateTime.UtcNow);
        }
        return reachable;
    }

    /// <summary>Backend resolution: LM Studio (local or remote Mac Mini) first,
    /// then the embedded llama-server sidecar, else null (caller falls back to
    /// the raw transcript).</summary>
    public static async Task<(string BaseUrl, string Model)?> ResolveBackendAsync(
        AppConfig config, LlamaSidecar embedded)
    {
        if (await IsReachableAsync(config.LmStudioUrl))
        {
            return (config.LmStudioUrl, config.CleanupModel);
        }
        string? embeddedUrl = await embedded.EnsureRunningAsync();
        if (embeddedUrl != null)
        {
            return (embeddedUrl, "embedded");
        }
        return null;
    }
}

File: macos/Sources/PrivateWhisper/CleanupService.swift
import Foundation

/// Talks to LM Studio's OpenAI-compatible endpoint for the cleanup pass and
/// command-mode rewrites. Any failure (unreachable, timeout, bad response)
/// throws — the pipeline falls back to the raw transcript so dictation is
/// never lost.
struct CleanupService {
    var baseURL: String
    var model: String
    var timeout: TimeInterval

    /// Canonical prompt lives in shared/prompts/cleanup_prompt.txt (bundled
    /// into Resources by build_app.sh); the literal below is the fallback so
    /// `swift build` binaries keep working without the bundle.
    static let systemPrompt: String = {
        if let url = Bundle.main.url(forResource: "cleanup_prompt", withExtension: "txt"),
           let text = try? String(contentsOf: url, encoding: .utf8),
           !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return text.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return fallbackSystemPrompt
    }()

    private static let fallbackSystemPrompt = """
        You clean up dictated text. Rules:
        - Output ONLY the cleaned text. No preamble, no quotes, no commentary.
        - Keep the same language as the input (German stays German, French stays French, English stays English).
        - Remove filler words (um, äh, also, alors, you know), false starts, and repetitions.
        - If the speaker corrects themselves ("next Tuesday — no wait, Wednesday", "also nein, ich meine…", "enfin, je veux dire…"), keep ONLY the corrected version.
        - If the speaker enumerates items ("first… second…", "erstens… zweitens…", "premièrement…"), format them as a list, one item per line, each starting with "- ".
        - Fix punctuation, capitalization, and obvious transcription errors.
        - Preserve meaning, tone, names, numbers, and technical terms exactly.
        - Do not summarize, do not expand, do not translate.
        """

    enum CleanupError: Error, LocalizedError {
        case badURL
        case badResponse(String)
        case emptyResult

        var errorDescription: String? {
            switch self {
            case .badURL: return "Invalid LM Studio URL"
            case .badResponse(let detail): return "LM Studio error: \(detail)"
            case .emptyResult: return "LM Studio returned an empty result"
            }
        }
    }

    func cleanup(
        transcript: String, language: String?,
        glossary: [String] = [], toneHint: String? = nil
    ) async throws -> String {
        var systemPrompt = Self.systemPrompt
        if let language, language != "unknown" {
            systemPrompt += "\n- The input language is \"\(language)\". The output must be in that same language."
        }
        if !glossary.isEmpty {
            systemPrompt += "\n- Personal dictionary — when the transcript contains a similar-sounding or misspelled variant of one of these, use this exact spelling: "
                + glossary.joined(separator: ", ")
        }
        if let toneHint, !toneHint.isEmpty {
            systemPrompt += "\n- The text will be inserted into an app where the expected style is: \(toneHint). Adjust register lightly; never change meaning."
        }

        // PRD: max_tokens sized to input length × 1.5 (≈3 chars/token heuristic),
        // plus fixed headroom for reasoning models that "think" before answering —
        // otherwise they hit the cap mid-thought and return empty content.
        // Non-thinking models just stop early, so the headroom is free.
        let maxTokens = max(256, Int(Double(transcript.count) / 3.0 * 1.5) + 64) + 2048

        return try await send(systemPrompt: systemPrompt, userContent: transcript, maxTokens: maxTokens)
    }

    /// Command mode: apply a spoken instruction to selected text.
    func rewrite(selection: String, instruction: String) async throws -> String {
        let systemPrompt: String
        if let url = Bundle.main.url(forResource: "rewrite_prompt", withExtension: "txt"),
           let text = try? String(contentsOf: url, encoding: .utf8),
           !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            systemPrompt = text.trimmingCharacters(in: .whitespacesAndNewlines)
        } else {
            systemPrompt = """
                You edit text according to the user's instruction. Rules:
                - Output ONLY the edited text. No preamble, no quotes, no commentary.
                - Keep the text's original language unless the instruction explicitly asks to translate.
                - Apply the instruction faithfully; change nothing else.
                """
        }
        let user = "Instruction: \(instruction)\n\nText:\n\(selection)"
        let maxTokens = max(512, Int(Double(selection.count) / 3.0 * 2)) + 2048
        return try await send(systemPrompt: systemPrompt, userContent: user, maxTokens: maxTokens)
    }

    private func send(systemPrompt: String, userContent: String, maxTokens: Int) async throws -> String {
        guard let url = URL(string: baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            .appending("/chat/completions")) else { throw CleanupError.badURL }

        var systemPrompt = systemPrompt
        // Qwen3 hybrids honor /no_think in the prompt; belt-and-braces next to
        // reasoning_effort below. Harmless elsewhere.
        if model.localizedCaseInsensitiveContains("qwen") {
            systemPrompt += "\n/no_think"
        }

        let body: [String: Any] = [
            "model": model,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user", "content": userContent],
            ],
            "temperature": 0.2,
            "max_tokens": maxTokens,
            "stream": false,
            // Disables thinking on models that support it (e.g. Qwen 3.5 —
            // 0.46s instead of 67s); safely ignored by non-reasoning models.
            "reasoning_effort": "none",
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.timeoutInterval = timeout

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            let detail = String(data: data.prefix(300), encoding: .utf8) ?? "HTTP error"
            throw CleanupError.badResponse(detail)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = json["choices"] as? [[String: Any]],
              let message = choices.first?["message"] as? [String: Any],
              let content = message["content"] as? String
        else {
            throw CleanupError.badResponse("unexpected response shape")
        }

        let cleaned = Self.stripReasoning(content).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else { throw CleanupError.emptyResult }
        return cleaned
    }

    /// Removes <think>…</think> blocks that reasoning models may emit.
    static func stripReasoning(_ text: String) -> String {
        var result = text.replacingOccurrences(
            of: #"(?s)<think>.*?</think>"#, with: "", options: .regularExpression)
        // Unterminated think block (hit max_tokens mid-reasoning) → nothing usable.
        if let range = result.range(of: "<think>") {
            result = String(result[..<range.lowerBound])
        }
        return result
    }

    /// Fetches available model IDs from LM Studio (for the settings dropdown).
    static func availableModels(baseURL: String) async -> [String] {
        guard let url = URL(string: baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            .appending("/models")) else { return [] }
        var request = URLRequest(url: url)
        request.timeoutInterval = 3
        guard let (data, _) = try? await URLSession.shared.data(for: request),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let models = json["data"] as? [[String: Any]]
        else { return [] }
        return models.compactMap { $0["id"] as? String }
    }
}
